dbms/src/Common/ColumnsHashing.h

```diff
@@ -570,6 +690,95 @@ struct HashMethodKeysFixed
     }
 };
 
+class KeySerializedBatchHandlerBase
+{
+private:
+    size_t batch_row_idx = 0;
+    String sort_key_container{};
+    PaddedPODArray<size_t> byte_size{};
+    PaddedPODArray<char *> pos{};
+    PaddedPODArray<char *> ori_pos{};
+    PaddedPODArray<size_t> real_byte_size{};
+
+    ALWAYS_INLINE inline void santityCheck() const
+    {
+        assert(ori_pos.size() == pos.size() && real_byte_size.size() == pos.size());
+    }
+
+    ALWAYS_INLINE inline void resize(size_t batch_size)
+    {
+        pos.resize(batch_size);
+        ori_pos.resize(batch_size);
+        real_byte_size.resize(batch_size);
+    }
+
+protected:
+    bool inited() const { return !byte_size.empty(); }
+
+    void init(size_t start_row, const ColumnRawPtrs & key_columns, const TiDB::TiDBCollators & collators)
+    {
+        batch_row_idx = start_row;
+        byte_size.resize_fill_zero(key_columns[0]->size());
+        RUNTIME_CHECK(!byte_size.empty());
+        for (size_t i = 0; i < key_columns.size(); ++i)
+            key_columns[i]->countSerializeByteSizeForCmp(
```

Comment:
From the following code, each non-null, non-collation string column will add sizeof(UINT32) to byte_size, not sure for multiple string columns, do we need to add this for each?
https://github.com/pingcap/tiflash/blob/94402d5f6dc3b883bc8fae544def6f19ec14df56/dbms/src/Columns/ColumnString.cpp#L582